### DISCUSSION TITLE
Updates from koord.live - including KoordASIO (again :)

### DISCUSSION THREAD

**Discussion by danryu** [jamulussoftware/jamulus#2914] [Ideas]
Hello to the global Jamulus community from Koord!

I just wanted to announce a little news from [Koord.Live](https://koord.live) that may interest Jamulus devs and users, and some upcoming changes that might be of PR-interest to upstream.

We have a [few PRs open already](https://github.com/jamulussoftware/jamulus/pulls/danryu) but are looking to make more significant contributions in the future. We are always wanting to push any critical bugfixes or improvements upstream, so please alert us here on our issues page if there's something you notice in our app that could be useful and that I haven't PR'd yet.

## KoordASIO
In particular I would like to mention the latest version of our open-source universal ASIO driver, KoordASIO - which is intended to wean the Windows world off the terrible affliction that is ASIO4ALL :) 
Github repo: https://github.com/koord-live/KoordASIO
Download from Microsoft Store: https://apps.microsoft.com/store/detail/koordasio-universal-driver/XP9CSS6NZBDV21

It includes the latest improvements from the upstream FlexASIO and PortAudio projects and provides the best and easiest way (IMO) for Windows users to use their generic hardware with ASIO, while benefiting from both WASAPI Shared and Exclusive Modes.
The GUI configurator is (I hope) intuitive and effective in setting safe, low-latency configurations.
Windows users, please give it a go and tell us what you think!
I previously raised this in an earlier discussion [here](https://github.com/jamulussoftware/jamulus/discussions/1936), and would be interested if Jamulus users and devs could report their experiences with it (and generate issues on our page if need be) - early testers are already coming back with positive stories!

As I mentioned previously, this could easily be integrated into Jamulus as an "internal ASIO" option in the dropdown list - with the same configuration options provided seamlessly in the Jamulus client as are in the KoordASIO control GUI.

## Koord app (our fork of Jamulus client)
Our Github repo here: https://github.com/koord-live/koord-app
Downloads page here: https://koord.live/downloads
Otherwise, on the client app side, these are the changes we've made at Koord which could be of interest to Jamulus devs (& users):
- now available in all Stores:
  - Apple App Store : macOS 11+ (intel and apple silicon) and iOS
    - also direct download notarized dmg installers for macOS
  - Google Play: Android 10+ (all ABIs supported)
  - Windows Store: Windows 10+, 64bit only (with bundled KoordASIO driver)
  - Linux AppImage - direct download - glibc 2.31+, 64-bit only 
- performance and stability fixes especially for mobile (Android + iOS)
- improvements in CI jobs for store submission
- modified UI for simplicity and mobile use
- integrated Video with the Koord.Live service and Koord app 
  - this is _not_ built into the Jamulus server (we basically run the vanilla Jamulus server build) but is rather our own customized Jitsi service running on compute-optimized bare-metal infrastructure, provided by Equinix Metal in all our core regions

Of course, please give the Koord app a try for your private sessions (or for whatever you like) and feel free to bombard us with bug reports and improvement suggestions!

## Upcoming dev objectives
Koord app improvements planned for the short-to-medium term:
- move to QML and a modern unified design
  - requires removing all Widgets, basically - and removing all the display logic from cpp files. IMO this will massively improve the efficiency of both core "business logic" devs and UI designers - business logic and display logic will no longer be intertwined in the same .cpp files!
  - this should ideally be a fairly transparent PR for upstream, and make life easier for everybody! It's a one-time job to make the transition, after which the advantages for both devs and users are cumulative…
- move to pure cmake-based builds
- app build CI automation improvements 
- Windows .msi installer build and auto-signing
- automated Windows Store submission
- resolve iOS TestFlight submission crash issues
- Build client app also as VST/AU plugin - to allow seamless integration in DAWs
- In the service: video recording, and live audio-video broadcast (probably will be a premium-only feature for now due to the compute requirements)
- Other new features - suggestions welcome :)

## Updates on the Koord service side:
- video now provided in all Regions and all session sizes, using the Koord app
- performance and stability fixes
- improved recordings management

Please feel free to check out our [free service](https://koord.live) and see what you think!
Suggestions and feedback always welcome.

**Comment by ann0see** [comment]
Thanks. I'll comment later on today or tomorrow. Just as quick reminder, the Jamulus team is not related to KOORD and I assume we both strive to be independent but cooperative meaning we will not advertise KOORD in preference to e.g Melomax

**Reply by danryu** [reply]
@ann0see Yes the intention was really to talk about the client-related developments on our side, and how that may possibly help upstream Jamulus in the future. I thought it was better to do an overview of our development, so that we are not working in disconnected silos.
I only mention the service in the context of these developments - I'm happy to remove those references if you prefer.
I don't know of any other groups doing development on this scale on active forks of Jamulus (client or server) so I didn't have a precedent to follow :)

I'm particularly interested if there is any renewed interest in KoordASIO/FlexASIO for use with Jamulus - this is obviously a generic universal ASIO driver utility with no necessary relation to the Koord service or even Jamulus.

I think Jamulus is a perfect use-case for KoordASIO/FlexASIO but YMMV.

**Reply by pljones** [reply]
I'd rather we had no ASIO support in Jamulus, to be quite honest, but we have to live in the current world where Windows is still stuck with it as the main (only real) profession driver API.  Even with JACK, you just make the ASIO driver JACK's problem rather than making it go away (JACK with PortAudio support for WASAPI, I guess, being next best).  Still, all an integrated solution like this does is replace "make it JACK's problem" with "make it our problem".  It doesn't make the problem go away.

It's still better to get a decent audio interface with a decent ASIO driver.  Encouraging other behaviour isn't, in my view, something we should work towards.

**Reply by danryu** [reply]
How does JACK make things easier or better for Windows users? 

If you’re stuck with ASIO on Windows (and you are) isn’t it better to include an integrated open-source fallback option with proven performance and flexibility, rather than design the installer and documentation around a poorly-maintained closed-source fallback?

**Reply by ann0see** [reply]
Native port audio would still be superior. There’s a feature branch which still needs some work. 

The last time I tested KoordAsio there was an error (but that was on an old laptop), but if I have time I‘ll try again on a fairly recent machine.

**Reply by dcorson-ticino-com** [reply]
I just tried this KoordAsio on my Win10 compi.  Note that ASIO4ALL works a little but crashes the compi after a couple of minutes (even when not using sound).
KoordAsio is not crashing, but adds an uncontrolled rhythm section to all sound.
Also not usable.
A native ASIO driver is the way to go.

**Reply by danryu** [reply]
> Native port audio would still be superior. 

Possibly but that's not a dead cert, I would say. It depends on what Windows audio backend is deployed by PortAudio. If it is WASAPI, then the feature branch will likely run into any problems as evidenced by KoordASIO/FlexASIO - the author of FlexASIO (which KoordASIO clones plus added GUI) talks at length about issues in the PortAudio WASAPI code that need addressing in the FlexASIO issues.

>  A native ASIO driver is the way to go.

This is a given. A universal driver can only be a fallback (though I suspect pretty good performance with KoordASIO and USB-audio devices - more so than with generic chipsets in most laptops - just a hunch). 
The point was really just to suggest KoordASIO as a viable open-source alternative to ASIO4ALL, if that is still being relied on by any poor souls :)

**Reply by ann0see** [reply]
@danryu just a small update here: Koord ASIO seems to work (while ASIO4ALL doesn't easily on my new laptop). But latency is quite high.

**Reply by danryu** [reply]
Thanks for the feedback @ann0see. Latency should be lower in Exclusive mode than in Shared, as per WASAPI design - though due to the current PortAudio/WASAPI situation I mentioned before, you may get some odd latency reporting and/or glitchiness.
Also useful to know would be the audio chipset in the laptop, as a lot hinges on that. Is it a RealTek HD generic as is so common these days?

**Reply by ann0see** [reply]
Yes, I used exclusive mode on a Realtek card. But latency is still quite high even with low buffer sizes.

**Reply by danryu** [reply]
I also do not have great results using my Realtek HD chipset in my Thinkpad.

The good thing is that KoordASIO performance/latency will automatically improve as PortAudio WASAPI support improves (which will take some effort but is feasible).

You can see some discussion with the FlexASIO author on this subject [here](https://github.com/dechamps/FlexASIO/issues/153#).

**Comment by ann0see** [comment]
> move to QML and a modern unified design

Yes. That would be a great improvement. 
We have internal design drafts.
